src/Mod/Sketcher/Gui/CommandSketcherTools.cpp

```diff
@@ -590,6 +590,58 @@ bool CmdSketcherSelectMalformedConstraints::isActive(void)
     return isSketcherAcceleratorActive(getActiveGuiDocument(), false);
 }
 
+DEF_STD_CMD_A(CmdSketcherSelectPartiallyRedundantConstraints)
+
+CmdSketcherSelectPartiallyRedundantConstraints::CmdSketcherSelectPartiallyRedundantConstraints()
+    :Command("Sketcher_SelectPartiallyRedundantConstraints")
+{
+    sAppModule      = "Sketcher";
+    sGroup          = QT_TR_NOOP("Sketcher");
+    sMenuText       = QT_TR_NOOP("Select partially redundant constraints");
+    sToolTipText    = QT_TR_NOOP("Select partially redundant constraints");
+    sWhatsThis      = "Sketcher_SelectPartiallyRedundantConstraints";
+    sStatusTip      = sToolTipText;
+    sPixmap         = "Sketcher_SelectPartiallyRedundantConstraints";
+    sAccel          = "CTRL+SHIFT+R";
+    eType           = ForEdit;
+}
+
+void CmdSketcherSelectPartiallyRedundantConstraints::activated(int iMsg)
+{
+    Q_UNUSED(iMsg);
+    Gui::Document * doc= getActiveGuiDocument();
+    ReleaseHandler(doc);
+    SketcherGui::ViewProviderSketch* vp = static_cast<SketcherGui::ViewProviderSketch*>(doc->getInEdit());
+    Sketcher::SketchObject* Obj= vp->getSketchObject();
+
+    std::string doc_name = Obj->getDocument()->getName();
+    std::string obj_name = Obj->getNameInDocument();
+
+    // get the needed lists and objects
+    const std::vector< int > &solverpartiallyredundant = vp->getSketchObject()->getLastPartiallyRedundant();
+    const std::vector< Sketcher::Constraint * > &vals = Obj->Constraints.getValues();
+
+    getSelection().clearSelection();
+
+    // push the constraints
+    int i = 0;
+    for (std::vector< Sketcher::Constraint * >::const_iterator it= vals.begin();it != vals.end(); ++it,++i) {
+        for(std::vector< int >::const_iterator itc= solverpartiallyredundant.begin();itc != solverpartiallyredundant.end(); ++itc) {
```

Comment:
```suggestion
        for(int partial : solverpartiallyredundant) {
```

Reply:
Thanks for reviewing. I will extend in this first comment beyond what is suggested to give you the rationale about why I tend to maintain style and reuse proven code at first instead of directly refactoring and deferring it.

Iterator based loop is equivalent to "new" (c++11) range based loop. There is no execution impact. Developers are used to it. One could talk about readability (range loop is generally more readable). But these files are full of iterator-based loops and it makes no sense to change the style of new code while leaving the rest (if one should be changed, all should, but iterators are still amply used in FreeCAD). I mostly tend to maintain the style of the existing files for new code. Obviously, the body of the loop is still expecting an iterator, so I reply to the intent rather than the fact that the code will not compile.

However, this file has a lot of analogous commands that are begging for a refactor. There are 3 or so "equal" commands save for if it is redundant, partially redundant or conflicting, such code repetition makes no sense and affects maintainability. In fact, the refactor should not just restrict to this file, but to  all the "chain" from GCS to the Viewer. All these are constraint ids (called Tags in the solver), it is simply the status that is different (redundant, conflicting,...). Therefore, one new class should unify this information and simplify the signaling mechanism. This is also desirable if the interface with the solver is to support future solver's, such as DeepSOIC's ConstraintSolver.

There are reasons why I do not refactor this at this time and prefer to replicate the existing structure. The first has to do with producing incremental code for new functionality that may need changes based on community feedback (it could even be removal of the feature and substitution with an alternative idea). The idea is to perform changes by addition, not by replacement. When the functionality settles down, then I think is a better time to consolidate and refactor. Unlike other projects where feedback comes from the customer after release, we work with very short latency in the feedback-implementation loop. This afternoon we will have the first impressions of this work and it is likely there will be change requests. The second relates to the release coming soon and not wanting to extend testing forever. Incremental code allows incremental testing. So I tend to produce the smallest differences, if possible reusing already tested code from other parts.

All that say, when v0.20 development cycle begins, it would be an appropriate time to refactor this. I have not forgotten about your other proposal either. If you identify parts that would benefit from refactoring, I am for refactoring.

Reply:
Thank you for taking the time  to respond :)